language-server/src/main/scala/scala/meta/languageserver/Compiler.scala

```diff
@@ -110,4 +127,22 @@ class Compiler(
     i + column
   }
 
+  private def ask[A](f: Response[A] => Unit): Response[A] = {
+    val r = new Response[A]
+    f(r)
+    r
+  }
+
+  private def typeOfTree(c: Global)(t: c.Tree): Option[String] = {
```

Comment:
Does `g.showCode(Tree)` give any better results? I recall it handling some cases a bit better while implementing the semanticdb-scalac type printer https://github.com/scalameta/scalameta/blob/c0b7b7df565115b39e0c1f6b6d39d307bef0b4ed/scalameta/semanticdb-scalac/src/main/scala/scala/meta/internal/semanticdb/PrinterOps.scala#L350-L458